File: DSA/searching_algorithms/binary_search/python/binary_search.py
#!/bin/env python3

"""
    Implementation of Linear Search Algorithm.
"""

def binary_search(arr, target):
    """Performs a binary search for target in arr. Returns index or -1 if not found."""
    start = 0 # Pointer for the First Element Index.
    end = len(arr) - 1 # Pointer for the Last Element Index.

    while start <= end:
        mid = start + (end - start) // 2 # Pointer for the Middle Element of Index.

        if arr[mid] == target:
            return mid
        elif target > arr[mid]:
            start = mid + 1
        else:
            end = mid - 1

    return -1

File: projects/FEM-DSA/search/binarySearch.js
#!/usr/bin/env node
/**
 * Implementation of Binary Search Algorithm.
 */

/**
 *
 * @param {*} arr : number[] -> Array of Numbers to Look within it.
 * @param {*} value : number -> The Value which we are looking for.
 * @returns {number} : The index of the value or -1;
 */
function binarySearch(arr, value) {
  let startIndex = 0;
  let endIndex = arr.length - 1;

  while (startIndex <= endIndex) {
    let midIndex = Math.floor((startIndex + endIndex) / 2);
    if (value === arr[midIndex]) {
      return midIndex;
    } else if (value > arr[midIndex]) {
      startIndex = midIndex + 1;
    } else {
      endIndex = midIndex - 1;
    }
  }

  return -1;
}

module.exports = binarySearch;
